src/NHSD.BuyingCatalogue.Identity.Api/Infrastructure/ValidationInputTagHelper.cs

```diff
@@ -0,0 +1,108 @@
+
+        [HtmlAttributeName(FieldDataTestIdName)]
+        public string FieldDataTestId { get; set; }
+
+        [HtmlAttributeName(InputDataTestIdName)]
+        public string InputDataTestId { get; set; }
+
+        [HtmlAttributeName(ErrorDataTestIdName)]
+        public string ErrorDataTestId { get; set; }
+
+        public override void Process(TagHelperContext context, TagHelperOutput output)
+        {
+            output.ThrowIfNull();
+
+            output.Content.Clear();
+
+            var outerDivBuilder = new TagBuilder(TagHelperConstants.Div);
+
+            if (ViewContext.ViewData.ModelState.ContainsKey(For.Name))
+            {
+                outerDivBuilder.AddCssClass(TagHelperConstants.NhsFormGroupError);
+            }
+
+            outerDivBuilder.Attributes[TagHelperConstants.DataTestId] = FieldDataTestId ?? $"{For.Name}-field";
+
+            var formGroup = new TagBuilder(TagHelperConstants.Div);
+            formGroup.AddCssClass(TagHelperConstants.NhsFormGroup);
+
+            var labelBuilder = _htmlGenerator.GenerateLabel(
+                ViewContext,
+                For.ModelExplorer,
+                For.Name,
+                null, 
+                null);
+
+            labelBuilder.AddCssClass(TagHelperConstants.NhsLabel);
+
+            var validationBuilder = _htmlGenerator.GenerateValidationMessage(ViewContext,
+                For.ModelExplorer,
+                For.Name,
+                null,
+                TagHelperConstants.Span,
+                null);
+
+            validationBuilder.AddCssClass(TagHelperConstants.NhsErrorMessage);
+            validationBuilder.Attributes[TagHelperConstants.DataTestId] = ErrorDataTestId ?? $"{For.Name}-error";
+            
+            var inputBuilder = _htmlGenerator.GenerateTextBox(ViewContext,
+                For.ModelExplorer,
+                For.Name,
+                null,
+                null,
+                null);
+
+            var dataTypeAttributes = For?.Metadata?
+                    .ContainerType?
+                    .GetProperty(For.Name)?
+                    .GetCustomAttributes<DataTypeAttribute>();
+            
+            if (dataTypeAttributes?.Any(x => x.DataType == DataType.Password) == true)
+            {
+                inputBuilder.Attributes[TagHelperConstants.Type] = "password";
+            }
+
+            inputBuilder.AddCssClass(TagHelperConstants.NhsInput);
+            inputBuilder.Attributes[TagHelperConstants.DataTestId] = InputDataTestId ?? $"{For.Name}-input";
+
+            formGroup.InnerHtml.AppendHtml(labelBuilder);
+            formGroup.InnerHtml.AppendHtml(validationBuilder);
+            formGroup.InnerHtml.AppendHtml(inputBuilder);
+            outerDivBuilder.InnerHtml.AppendHtml(formGroup);
+
+            output.TagName = TagHelperConstants.Div;
+            output.TagMode = TagMode.StartTagAndEndTag;
+            output.Content.AppendHtml(outerDivBuilder);
+        }
```

Comment:
Nice. Might it be worth moving the build of some of the inner elements out into separate methods, e.g. label, validation, input etc? (Purely for readability.)